diff --git a/dist/types/Delta.d.ts b/dist/types/Delta.d.ts
index 50df51f13a490ee065067b0a5b045a42370886c6..88ecefafd29b8643d3c39bae85787113a8e2a7e9 100644
--- a/dist/types/Delta.d.ts
+++ b/dist/types/Delta.d.ts
@@ -1,7 +1,7 @@
 import diff from 'fast-diff';
-import AttributeMap from './AttributeMap';
-import Op from './Op';
-import OpIterator from './OpIterator';
+import AttributeMap from './AttributeMap.js';
+import Op from './Op.js';
+import OpIterator from './OpIterator.js';
 interface EmbedHandler<T> {
     compose(a: T, b: T, keepNull: boolean): T;
     invert(a: T, b: T): T;
diff --git a/dist/types/Op.d.ts b/dist/types/Op.d.ts
index fc469f835731a86fbb893047f446beba5f7d864a..7189aca98a4fbf53d1aaafc75e22f0e5fa07694a 100644
--- a/dist/types/Op.d.ts
+++ b/dist/types/Op.d.ts
@@ -1,4 +1,4 @@
-import AttributeMap from './AttributeMap';
+import AttributeMap from './AttributeMap.js';
 interface Op {
     insert?: string | Record<string, unknown>;
     delete?: number;
diff --git a/dist/types/OpIterator.d.ts b/dist/types/OpIterator.d.ts
index aa8de0844f2da825cee54424086722ec619bbba9..1b8da3ca2772c33b1fbcbd8e2b3f9eb08e782e34 100644
--- a/dist/types/OpIterator.d.ts
+++ b/dist/types/OpIterator.d.ts
@@ -1,4 +1,4 @@
-import Op from './Op';
+import Op from './Op.js';
 export default class Iterator {
     ops: Op[];
     index: number;
diff --git a/dist/types/package.json b/dist/types/package.json
new file mode 100644
index 0000000000000000000000000000000000000000..3dbc1ca591c0557e35b6004aeba250e6a70b56e3
--- /dev/null
+++ b/dist/types/package.json
@@ -0,0 +1,3 @@
+{
+  "type": "module"
+}
