diff --git a/database.d.ts b/database.d.ts
index 0df6d67e6c13151bfec5279fe72678b87893abb1..6a77fca080e94a3fa1291540788ef94118fba732 100644
--- a/database.d.ts
+++ b/database.d.ts
@@ -1,4 +1,4 @@
-import { Emoji, DatabaseConstructorOptions, SkinTone, CustomEmoji, NativeEmoji } from "./shared";
+import { Emoji, DatabaseConstructorOptions, SkinTone, CustomEmoji, NativeEmoji } from "./shared.js";
 export default class Database {
     /**
      * Create a new Database.
diff --git a/index.d.ts b/index.d.ts
index 978bd985307cd8acdb1c063200501cc6615ff0b0..f331e68964714e81693c195c21f3cbed44ab1c32 100644
--- a/index.d.ts
+++ b/index.d.ts
@@ -1,5 +1,5 @@
-import Picker from './picker'
-import Database from './database'
+import Picker from './picker.js'
+import Database from './database.js'
 
 export {
   Picker,
diff --git a/picker.d.ts b/picker.d.ts
index e28707250358ac2310484b16876e0675dc2a1593..7bb11d13dc59bb591cf3b405c5c1190faac7483f 100644
--- a/picker.d.ts
+++ b/picker.d.ts
@@ -1,4 +1,4 @@
-import { I18n, PickerConstructorOptions, EmojiPickerEventMap, CustomEmoji } from "./shared";
+import { I18n, PickerConstructorOptions, EmojiPickerEventMap, CustomEmoji } from "./shared.js";
 export default class Picker extends HTMLElement {
     dataSource: string;
     locale: string;
diff --git a/trimEmojiData.d.ts b/trimEmojiData.d.ts
index 312228a373671da1d1916b93fe3ccf50a271f285..7190a01ba0285c956346cda078a1c57d6b76bfac 100644
--- a/trimEmojiData.d.ts
+++ b/trimEmojiData.d.ts
@@ -1,4 +1,4 @@
-import { Emoji } from "./shared";
+import { Emoji } from "./shared.js";
 /**
  * Given an array of emoji (e.g. emojibase-data/en/data.json), extract out only the data
  * needed for emoji-picker-element, resulting in a smaller object.
