diff --git a/dist/es6/jssm.d.ts b/dist/es6/jssm.d.ts
index 26da4815c090031aaeb8f367425503dc8d2a323d..1eaa56168048b46fc2e6e6dc5ade7a8d5151db8a 100644
--- a/dist/es6/jssm.d.ts
+++ b/dist/es6/jssm.d.ts
@@ -1,12 +1,12 @@
 declare type StateType = string;
 import { JssmGenericState, JssmGenericConfig, JssmStateConfig, JssmTransition, JssmTransitionList, // JssmTransitionRule,
-JssmMachineInternalState, JssmAllowsOverride, JssmStateDeclaration, JssmStateStyleKeyList, JssmLayout, JssmHistory, JssmSerialization, FslDirection, FslDirections, FslTheme, HookDescription, HookHandler, HookContext, HookResult, HookComplexResult, JssmRng } from './jssm_types';
-import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow';
-import { compile, make, wrap_parse } from './jssm_compiler';
-import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key } from './jssm_util';
-import * as constants from './jssm_constants';
+JssmMachineInternalState, JssmAllowsOverride, JssmStateDeclaration, JssmStateStyleKeyList, JssmLayout, JssmHistory, JssmSerialization, FslDirection, FslDirections, FslTheme, HookDescription, HookHandler, HookContext, HookResult, HookComplexResult, JssmRng } from './jssm_types.js';
+import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow.js';
+import { compile, make, wrap_parse } from './jssm_compiler.js';
+import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key } from './jssm_util.js';
+import * as constants from './jssm_constants.js';
 declare const shapes: string[], gviz_shapes: string[], named_colors: string[];
-import { version, build_time } from './version';
+import { version, build_time } from './version.js';
 /*********
  *
  *  An internal method meant to take a series of declarations and fold them into
diff --git a/dist/es6/jssm.js b/dist/es6/jssm.js
index 679e3c62c5b5b6a7c182672a2573321c090c92dc..30224bc19be5422490db105d1449f381e103b231 100644
--- a/dist/es6/jssm.js
+++ b/dist/es6/jssm.js
@@ -1,14 +1,14 @@
 // whargarbl lots of these return arrays could/should be sets
 import { circular_buffer } from 'circular_buffer_js';
-import { FslDirections } from './jssm_types';
-import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow';
-import { compile, make, wrap_parse } from './jssm_compiler';
-import { theme_mapping, base_theme } from './jssm_theme';
-import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key, array_box_if_string, name_bind_prop_and_state, hook_name, named_hook_name, gen_splitmix32 } from './jssm_util';
-import * as constants from './jssm_constants';
+import { FslDirections } from './jssm_types.js';
+import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow.js';
+import { compile, make, wrap_parse } from './jssm_compiler.js';
+import { theme_mapping, base_theme } from './jssm_theme.js';
+import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key, array_box_if_string, name_bind_prop_and_state, hook_name, named_hook_name, gen_splitmix32 } from './jssm_util.js';
+import * as constants from './jssm_constants.js';
 const { shapes, gviz_shapes, named_colors } = constants;
-import { version, build_time } from './version'; // replaced from package.js in build
-import { JssmError } from './jssm_error';
+import { version, build_time } from './version.js'; // replaced from package.js in build
+import { JssmError } from './jssm_error.js';
 /*********
  *
  *  An internal method meant to take a series of declarations and fold them into
diff --git a/dist/es6/jssm_arrow.d.ts b/dist/es6/jssm_arrow.d.ts
index 3d56083828c95afd94d52dc423277fdebe5cdaf3..5d5635717a3fef9c75ffbe32258cf39a4729c6e2 100644
--- a/dist/es6/jssm_arrow.d.ts
+++ b/dist/es6/jssm_arrow.d.ts
@@ -1,4 +1,4 @@
-import { JssmArrow, JssmArrowDirection, JssmArrowKind } from './jssm_types';
+import { JssmArrow, JssmArrowDirection, JssmArrowKind } from './jssm_types.js';
 /*********
  *
  *  Return the direction of an arrow - `right`, `left`, or `both`.
diff --git a/dist/es6/jssm_arrow.js b/dist/es6/jssm_arrow.js
index e4ae26bcadddd7b593a80f4c132fc3c3fa27ab9a..caea9cf116212602ee12d912a29ccdceb5f9daff 100644
--- a/dist/es6/jssm_arrow.js
+++ b/dist/es6/jssm_arrow.js
@@ -1,4 +1,4 @@
-import { JssmError } from './jssm_error';
+import { JssmError } from './jssm_error.js';
 /* eslint-disable complexity */
 /*********
  *
diff --git a/dist/es6/jssm_compiler.d.ts b/dist/es6/jssm_compiler.d.ts
index 2bf6c2594c8d285ac5dd768b609242f2d3581779..5db03761aa694ba4d54b1d925a0ad3ced8bcab32 100644
--- a/dist/es6/jssm_compiler.d.ts
+++ b/dist/es6/jssm_compiler.d.ts
@@ -1,4 +1,4 @@
-import { JssmTransition, JssmCompileSe, JssmParseTree, JssmGenericConfig } from './jssm_types';
+import { JssmTransition, JssmCompileSe, JssmParseTree, JssmGenericConfig } from './jssm_types.js';
 /*********
  *
  *  Internal method meant to perform factory assembly of an edge.  Not meant for
diff --git a/dist/es6/jssm_compiler.js b/dist/es6/jssm_compiler.js
index 934672762b5a465a4433a8a578575855d4fd4624..5622289875de94a5517a62e4169bb9e4f6246854 100644
--- a/dist/es6/jssm_compiler.js
+++ b/dist/es6/jssm_compiler.js
@@ -1,7 +1,7 @@
-import { JssmError } from './jssm_error';
-import { parse } from './fsl_parser';
-import { arrow_left_kind, arrow_right_kind } from './jssm_arrow';
-import { find_repeated, name_bind_prop_and_state } from './jssm_util';
+import { JssmError } from './jssm_error.js';
+import { parse } from './fsl_parser.js';
+import { arrow_left_kind, arrow_right_kind } from './jssm_arrow.js';
+import { find_repeated, name_bind_prop_and_state } from './jssm_util.js';
 import { reduce as reduce_to_639 } from 'reduce-to-639-1';
 /*********
  *
diff --git a/dist/es6/jssm_error.d.ts b/dist/es6/jssm_error.d.ts
index 1232f995f924d0a04f06597ff01df9b3f9245e6d..54da23fc7a85185e02ee5349fae0e1ed37b2dd49 100644
--- a/dist/es6/jssm_error.d.ts
+++ b/dist/es6/jssm_error.d.ts
@@ -1,4 +1,4 @@
-import { JssmErrorExtendedInfo } from './jssm_types';
+import { JssmErrorExtendedInfo } from './jssm_types.js';
 declare class JssmError extends Error {
     message: string;
     base_message: string;
diff --git a/dist/es6/jssm_theme.d.ts b/dist/es6/jssm_theme.d.ts
index d5713932708d0b98aee381347ec854c656f3e5a1..630c91ed50d2d5bf4e80dcd04689bc3dfc76074b 100644
--- a/dist/es6/jssm_theme.d.ts
+++ b/dist/es6/jssm_theme.d.ts
@@ -1,4 +1,4 @@
-import { FslTheme, JssmBaseTheme } from './jssm_types';
-import { base_theme } from './themes/jssm_base_stylesheet';
+import { FslTheme, JssmBaseTheme } from './jssm_types.js';
+import { base_theme } from './themes/jssm_base_stylesheet.js';
 declare const theme_mapping: Map<FslTheme, JssmBaseTheme>;
 export { theme_mapping, base_theme };
diff --git a/dist/es6/jssm_theme.js b/dist/es6/jssm_theme.js
index 89fcef21bef03c9d755aac331411cabdab343955..b667417413e500a8d1a9e97dd2f351a060e49137 100644
--- a/dist/es6/jssm_theme.js
+++ b/dist/es6/jssm_theme.js
@@ -1,9 +1,9 @@
-import { base_theme } from './themes/jssm_base_stylesheet';
-import { default_theme } from './themes/jssm_theme_default';
-import { modern_theme } from './themes/jssm_theme_modern';
-import { ocean_theme } from './themes/jssm_theme_ocean';
-import { plain_theme } from './themes/jssm_theme_plain';
-import { bold_theme } from './themes/jssm_theme_bold';
+import { base_theme } from './themes/jssm_base_stylesheet.js';
+import { default_theme } from './themes/jssm_theme_default.js';
+import { modern_theme } from './themes/jssm_theme_modern.js';
+import { ocean_theme } from './themes/jssm_theme_ocean.js';
+import { plain_theme } from './themes/jssm_theme_plain.js';
+import { bold_theme } from './themes/jssm_theme_bold.js';
 const theme_mapping = new Map();
 theme_mapping.set('default', default_theme);
 theme_mapping.set('modern', modern_theme);
diff --git a/dist/es6/jssm_util.js b/dist/es6/jssm_util.js
index c5ba0cc981469ad1c1b8eede9cc705e179224a41..e5f0e90da339f42e04ad994633e1c45d9d439fd8 100644
--- a/dist/es6/jssm_util.js
+++ b/dist/es6/jssm_util.js
@@ -1,4 +1,4 @@
-import { JssmError } from './jssm_error';
+import { JssmError } from './jssm_error.js';
 /*******
  *
  *  Predicate for validating an array for uniqueness.  Not generally meant for
diff --git a/dist/es6/themes/jssm_base_stylesheet.d.ts b/dist/es6/themes/jssm_base_stylesheet.d.ts
index bcf0a491028364d6a3bed2506a14589d14d7cf44..5b32f5c109bedd2c6ab54ce8dd28b0ab2e199626 100644
--- a/dist/es6/themes/jssm_base_stylesheet.d.ts
+++ b/dist/es6/themes/jssm_base_stylesheet.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/themes/jssm_theme_bold.d.ts b/dist/es6/themes/jssm_theme_bold.d.ts
index d7cd1a56c1f3ca4fede96698cdfe3656f9dced59..0eccd944d45201218d6693e41c0c722caeacb9aa 100644
--- a/dist/es6/themes/jssm_theme_bold.d.ts
+++ b/dist/es6/themes/jssm_theme_bold.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/themes/jssm_theme_default.d.ts b/dist/es6/themes/jssm_theme_default.d.ts
index bade2da1e9559bf2b513e4fca1e0a6d3e822cbf0..1d6215957f28d252d6bc46384237e89920bdb393 100644
--- a/dist/es6/themes/jssm_theme_default.d.ts
+++ b/dist/es6/themes/jssm_theme_default.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/themes/jssm_theme_modern.d.ts b/dist/es6/themes/jssm_theme_modern.d.ts
index e0ecd4bf5059d9677adbe044ea51642ceab321f8..dfb6c3148110af98d0dc3d78d11f3afec90a1a8a 100644
--- a/dist/es6/themes/jssm_theme_modern.d.ts
+++ b/dist/es6/themes/jssm_theme_modern.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/themes/jssm_theme_ocean.d.ts b/dist/es6/themes/jssm_theme_ocean.d.ts
index d2db15fe93b46fe23a829219860640c65f3b1084..3fa6666ae2baf1ebda0c17a387acbe969eae1d66 100644
--- a/dist/es6/themes/jssm_theme_ocean.d.ts
+++ b/dist/es6/themes/jssm_theme_ocean.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/themes/jssm_theme_plain.d.ts b/dist/es6/themes/jssm_theme_plain.d.ts
index 0d1d71d8eb028df651490038dbed06b89fc82f4d..0ff4641de5fa587751f1c4afbd4600278d2a3bf4 100644
--- a/dist/es6/themes/jssm_theme_plain.d.ts
+++ b/dist/es6/themes/jssm_theme_plain.d.ts
@@ -1,4 +1,4 @@
-import { JssmStateConfig, JssmBaseTheme } from '../jssm_types';
+import { JssmStateConfig, JssmBaseTheme } from '../jssm_types.js';
 declare const base_state_style: JssmStateConfig;
 declare const base_active_state_style: JssmStateConfig;
 declare const base_terminal_state_style: JssmStateConfig;
diff --git a/dist/es6/version.js b/dist/es6/version.js
index 732d2603444e5cd2c08a90815770cb491d41aff3..3ea98720c0de66a27be8f8a517371805ed004b9f 100644
--- a/dist/es6/version.js
+++ b/dist/es6/version.js
@@ -1,2 +1,2 @@
-const version = "5.97.1", build_time = 1707173654610;
+const version = "5.97.1", build_time = 1709587899623;
 export { version, build_time };
diff --git a/jssm.d.ts b/jssm.d.ts
index 26da4815c090031aaeb8f367425503dc8d2a323d..1eaa56168048b46fc2e6e6dc5ade7a8d5151db8a 100644
--- a/jssm.d.ts
+++ b/jssm.d.ts
@@ -1,12 +1,12 @@
 declare type StateType = string;
 import { JssmGenericState, JssmGenericConfig, JssmStateConfig, JssmTransition, JssmTransitionList, // JssmTransitionRule,
-JssmMachineInternalState, JssmAllowsOverride, JssmStateDeclaration, JssmStateStyleKeyList, JssmLayout, JssmHistory, JssmSerialization, FslDirection, FslDirections, FslTheme, HookDescription, HookHandler, HookContext, HookResult, HookComplexResult, JssmRng } from './jssm_types';
-import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow';
-import { compile, make, wrap_parse } from './jssm_compiler';
-import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key } from './jssm_util';
-import * as constants from './jssm_constants';
+JssmMachineInternalState, JssmAllowsOverride, JssmStateDeclaration, JssmStateStyleKeyList, JssmLayout, JssmHistory, JssmSerialization, FslDirection, FslDirections, FslTheme, HookDescription, HookHandler, HookContext, HookResult, HookComplexResult, JssmRng } from './jssm_types.js';
+import { arrow_direction, arrow_left_kind, arrow_right_kind } from './jssm_arrow.js';
+import { compile, make, wrap_parse } from './jssm_compiler.js';
+import { seq, unique, find_repeated, weighted_rand_select, weighted_sample_select, histograph, weighted_histo_key } from './jssm_util.js';
+import * as constants from './jssm_constants.js';
 declare const shapes: string[], gviz_shapes: string[], named_colors: string[];
-import { version, build_time } from './version';
+import { version, build_time } from './version.js';
 /*********
  *
  *  An internal method meant to take a series of declarations and fold them into
diff --git a/jssm_arrow.d.ts b/jssm_arrow.d.ts
index 3d56083828c95afd94d52dc423277fdebe5cdaf3..5d5635717a3fef9c75ffbe32258cf39a4729c6e2 100644
--- a/jssm_arrow.d.ts
+++ b/jssm_arrow.d.ts
@@ -1,4 +1,4 @@
-import { JssmArrow, JssmArrowDirection, JssmArrowKind } from './jssm_types';
+import { JssmArrow, JssmArrowDirection, JssmArrowKind } from './jssm_types.js';
 /*********
  *
  *  Return the direction of an arrow - `right`, `left`, or `both`.
diff --git a/jssm_compiler.d.ts b/jssm_compiler.d.ts
index 2bf6c2594c8d285ac5dd768b609242f2d3581779..5db03761aa694ba4d54b1d925a0ad3ced8bcab32 100644
--- a/jssm_compiler.d.ts
+++ b/jssm_compiler.d.ts
@@ -1,4 +1,4 @@
-import { JssmTransition, JssmCompileSe, JssmParseTree, JssmGenericConfig } from './jssm_types';
+import { JssmTransition, JssmCompileSe, JssmParseTree, JssmGenericConfig } from './jssm_types.js';
 /*********
  *
  *  Internal method meant to perform factory assembly of an edge.  Not meant for
diff --git a/jssm_error.d.ts b/jssm_error.d.ts
index 1232f995f924d0a04f06597ff01df9b3f9245e6d..54da23fc7a85185e02ee5349fae0e1ed37b2dd49 100644
--- a/jssm_error.d.ts
+++ b/jssm_error.d.ts
@@ -1,4 +1,4 @@
-import { JssmErrorExtendedInfo } from './jssm_types';
+import { JssmErrorExtendedInfo } from './jssm_types.js';
 declare class JssmError extends Error {
     message: string;
     base_message: string;
diff --git a/jssm_theme.d.ts b/jssm_theme.d.ts
index d5713932708d0b98aee381347ec854c656f3e5a1..630c91ed50d2d5bf4e80dcd04689bc3dfc76074b 100644
--- a/jssm_theme.d.ts
+++ b/jssm_theme.d.ts
@@ -1,4 +1,4 @@
-import { FslTheme, JssmBaseTheme } from './jssm_types';
-import { base_theme } from './themes/jssm_base_stylesheet';
+import { FslTheme, JssmBaseTheme } from './jssm_types.js';
+import { base_theme } from './themes/jssm_base_stylesheet.js';
 declare const theme_mapping: Map<FslTheme, JssmBaseTheme>;
 export { theme_mapping, base_theme };
diff --git a/package.json b/package.json
index 1eb1f70fee424f5686471377b04e035ea5cbf2f4..b2559767e7d45c255de4a44f3e71c1024e3f7c4a 100644
--- a/package.json
+++ b/package.json
@@ -5,10 +5,22 @@
     "node": ">=10.0.0"
   },
   "exports": {
-    "require": "./dist/jssm.es5.cjs",
-    "import": "./dist/jssm.es6.mjs",
-    "default": "./dist/jssm.es5.cjs",
-    "browser": "dist/jssm.es5.iife.cjs"
+    "require": {
+      "default": "./dist/jssm.es5.cjs.js",
+      "types": "./jssm.d.ts"
+    },
+    "import": {
+      "default": "./dist/es6/jssm.js",
+      "types": "./jssm.d.ts"
+    },
+    "default": {
+      "default": "./dist/jssm.es5.cjs.js",
+      "types": "./jssm.d.ts"
+    },
+    "browser": {
+      "default": "./dist/jssm.es5.iife.js",
+      "types": "./jssm.d.ts"
+    }
   },
   "autoupdate": {
     "source": "git",
@@ -27,7 +39,6 @@
   "main": "dist/jssm.es5.cjs",
   "module": "dist/jssm.es6.js",
   "browser": "dist/jssm.es5.iife.cjs",
-  "types": "./jssm.d.ts",
   "scripts": {
     "jest-unicode-action": "jest unicode-actions.uspec.ts -c jest-unicode.config.cjs --color --verbose",
     "jest-unicode-string": "jest unicode-strings.uspec.ts -c jest-unicode.config.cjs --color --verbose",